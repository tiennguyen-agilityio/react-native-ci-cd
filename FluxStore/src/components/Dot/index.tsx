import {memo, useMemo} from 'react';
import Animated, {
  runOnJS,
  useAnimatedStyle,
  useSharedValue,
  withSpring,
} from 'react-native-reanimated';
import {Gesture, GestureDetector} from 'react-native-gesture-handler';

// Stores

// Themes
import {colors} from '@/themes';

// Components
import Flex from '../Flex';

interface DotProps {
  color: string;
  hasBorder: boolean;
  size?: number;
  onSelect: () => void;
}

const Dot = ({color, hasBorder, size = 34, onSelect}: DotProps) => {
  const scale = useSharedValue(1);

  const rStyle = useAnimatedStyle(() => ({
    transform: [{scale: scale.value}],
  }));

  const dotStyles = useMemo(() => {
    if (hasBorder) {
      return {
        width: size,
        height: size,
        borderRadius: '50%',
        borderColor: colors.white[500],
        borderWidth: 5,
        backgroundColor: color,
        shadowOffset: {width: 0, height: 3},
        shadowOpacity: 0.2,
        shadowRadius: 6,
        elevation: 4,
      };
    }

    return {
      width: size - 10,
      height: size - 10,
      borderRadius: '50%',
      shadowColor: colors.black[500],
      backgroundColor: color,
    };
  }, [color, hasBorder, size]);

  const handleSelect = () => {
    'worklet';

    scale.value = withSpring(1.1, {}, () => {
      scale.value = withSpring(1);
      runOnJS(onSelect)();
    });
  };

  const press = Gesture.Pan()
    .onBegin(handleSelect)
    ?.onEnd(() => {
      'worklet';
      scale.value = withSpring(1, {duration: 180});
    });

  const composed = Gesture.Simultaneous(press);

  return (
    <Flex width={size} height={size} align="center" justify="center">
      <GestureDetector gesture={composed}>
        <Animated.View testID={`dot-${color}`} style={[dotStyles, rStyle]} />
      </GestureDetector>
    </Flex>
  );
};

export default memo(Dot);
