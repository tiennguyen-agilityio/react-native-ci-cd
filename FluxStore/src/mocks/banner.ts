export const BANNERS = [
  {
    id: '01',
    title: 'Clothing Collection 2025',
    imageUrl: 'https://i.ibb.co/wZv2s1T6/Frame-33149.png',
  },
  {
    id: '02',
    title: 'Autumn Collection 2025',
    imageUrl: 'https://i.ibb.co/hRBFHvHk/Mask-Group-1.png',
  },
  {
    id: '03',
    title: 'Accessories Collection 2025',
    imageUrl: 'https://i.ibb.co/20WCrpXW/Frame-33123.png',
  },
];
