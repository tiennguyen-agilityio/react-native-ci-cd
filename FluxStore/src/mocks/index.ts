export * from './product';
export * from './banner';
export * from './review';
export * from './cart';
