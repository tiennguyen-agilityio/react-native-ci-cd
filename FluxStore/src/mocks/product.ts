import {Product} from '@/interfaces';

export const COLORS = ['black', 'red', 'blue'];
export const SIZES = ['S', 'M', 'L'];

export const PRODUCTS: Product[] = [
  {
    id: '01',
    name: 'Linen Dress',
    price: 90.02,
    discount: 10,
    rating: 5,
    reviewCount: 20,
    image: 'https://i.ibb.co/SD47BZNp/Mask-Group-1.png',
    description:
      'Minimal Stand is made of by natural wood. The design that is very simple and minimal. This is truly one of the best furnitures in any family for now. With 3 different colors, you can easily select the best match for your home.',
    sizes: ['s', 'M'],
    colors: ['black', 'red', 'blue'],
    carouselImages: [
      'https://i.ibb.co/KQj6DyL/image-87.png',
      'https://i.ibb.co/SD47BZNp/Mask-Group-1.png',
      'https://i.ibb.co/Jw9tjsj3/Mask-Group-2.png',
    ],
  },
  {
    id: '02',
    name: 'Filted Waist Dress',
    price: 82.0,
    discount: 0,
    rating: 3,
    reviewCount: 110,
    image: 'https://i.ibb.co/Jw9tjsj3/Mask-Group-2.png',
    description:
      'Minimal Stand is made of by natural wood. The design that is very simple and minimal. This is truly one of the best furnitures in any family for now. With 3 different colors, you can easily select the best match for your home.',
    sizes: ['XL'],
    colors: ['pink', 'black'],
    carouselImages: [
      'https://i.ibb.co/KQj6DyL/image-87.png',
      'https://i.ibb.co/SD47BZNp/Mask-Group-1.png',
      'https://i.ibb.co/Jw9tjsj3/Mask-Group-2.png',
    ],
  },
];
