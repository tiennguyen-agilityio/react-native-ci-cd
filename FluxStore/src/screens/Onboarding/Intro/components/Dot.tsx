import {colors} from '@/themes';
import {memo, useCallback} from 'react';
import {Pressable, StyleSheet} from 'react-native';
import Animated, {
  interpolateColor,
  SharedValue,
  useAnimatedStyle,
  withTiming,
} from 'react-native-reanimated';

const styles = StyleSheet.create({
  dot: {
    width: 6,
    height: 6,
    borderRadius: 3,
    borderWidth: 1,
    borderColor: colors.white[500],
  },
});

interface DotProps {
  index: number;
  currentIndex: SharedValue<number>;
  onPress: (index: number) => void;
}

const Dot = ({index, currentIndex, onPress}: DotProps) => {
  const animatedStyle = useAnimatedStyle(() => {
    const backgroundColor = interpolateColor(
      currentIndex.value,
      [index - 1, index, index + 1],
      [colors.transparent, colors.white[500], colors.transparent],
    );

    return {
      backgroundColor,
      transform: [
        {
          scale: withTiming(currentIndex.value === index ? 1.2 : 1),
        },
      ],
    };
  });

  const handlePress = useCallback(() => {
    onPress(index);
  }, [index, onPress]);

  return (
    <Pressable onPress={handlePress}>
      <Animated.View style={[styles.dot, animatedStyle]} />
    </Pressable>
  );
};

export default memo(Dot);
