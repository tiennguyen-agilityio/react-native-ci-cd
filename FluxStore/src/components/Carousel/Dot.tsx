import {Flex} from '@/components';
import {colors} from '@/themes';
import {memo, useCallback, useMemo} from 'react';
import {Pressable, StyleSheet} from 'react-native';
import Animated, {SharedValue, useAnimatedStyle, withTiming} from 'react-native-reanimated';

interface DotProps {
  index: number;
  color?: string;
  currentIndex: SharedValue<number>;
  onPress: (index: number) => void;
}

const Dot = ({index, currentIndex, color, onPress}: DotProps) => {
  const animatedStyle = useAnimatedStyle(() => {
    const isCurrentValue = currentIndex.value === index;

    return {
      backgroundColor: colors.transparent,
      borderWidth: withTiming(isCurrentValue ? 1 : 0),
      borderColor: withTiming(isCurrentValue ? color ?? colors.white[500] : colors.transparent),
      transform: [
        {
          scale: withTiming(isCurrentValue ? 1.3 : 1),
        },
      ],
    };
  });

  const styles = useMemo(
    () =>
      StyleSheet.create({
        wrapper: {
          width: 10.5,
          height: 10.5,
          borderRadius: 10.5,
          borderWidth: 3,
          padding: 1.5,
          justifyContent: 'center',
          overflow: 'hidden',
          alignItems: 'center',
          backgroundColor: colors.transparent,
          borderColor: color ?? colors.white[500],
        },
        dot: {
          width: 4.5,
          height: 4.5,
          borderRadius: 4.5,
          overflow: 'hidden',
          backgroundColor: color ?? colors.white[500],
        },
      }),
    [color],
  );

  const handlePress = useCallback(() => {
    onPress(index);
  }, [index, onPress]);

  return (
    <Pressable onPress={handlePress}>
      <Animated.View style={[styles.wrapper, animatedStyle]}>
        <Flex style={styles.dot} />
      </Animated.View>
    </Pressable>
  );
};

export default memo(Dot);
