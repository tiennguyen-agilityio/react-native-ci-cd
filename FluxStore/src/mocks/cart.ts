import {PRODUCTS} from './product';

export const CARTS = [
  {
    id: '01',
    product: {
      id: PRODUCTS[0].id,
      image: PRODUCTS[0].image,
      name: PRODUCTS[0].name,
      price: PRODUCTS[0].price,
    },
    sizes: [PRODUCTS[0]?.sizes?.[0], PRODUCTS[0]?.sizes?.[1]].join(', '),
    colors: [PRODUCTS[0]?.colors?.[0]].join(', '),
    quantity: 2,
    isChecked: true,
  },
  {
    id: '02',
    product: {
      id: PRODUCTS[1].id,
      image: PRODUCTS[1].image,
      name: PRODUCTS[1].name,
      price: PRODUCTS[1].price,
    },
    sizes: [PRODUCTS[1]?.sizes?.[0]].join(', '),
    colors: [PRODUCTS[1]?.colors?.[0], PRODUCTS[1]?.colors?.[1]].join(', '),
    quantity: 1,
    isChecked: false,
  },
];
