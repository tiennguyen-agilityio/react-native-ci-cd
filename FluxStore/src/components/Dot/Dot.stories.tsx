import {Meta, StoryObj} from '@storybook/react';

import {COLORS} from '@/mocks';

import Dot from '.';
import {useState} from 'react';

const meta: Meta<typeof Dot> = {
  title: 'Components/Dot',
  component: Dot,
};

export default meta;

type Story = StoryObj<typeof Dot>;

const defaultProps = {
  color: COLORS[0],
  hasBorder: false,
  size: 34,
};

const Render = (props: {hasBorder?: boolean}) => {
  const [isShowBorder, setIsShowBorder] = useState(false);
  const handleChange = () => setIsShowBorder(prev => !prev);

  return <Dot {...defaultProps} hasBorder={isShowBorder} onSelect={handleChange} {...props} />;
};

export const Default: Story = {
  render: Render,
};

export const HasBorder: Story = {
  render: () => Render({hasBorder: true}),
};
