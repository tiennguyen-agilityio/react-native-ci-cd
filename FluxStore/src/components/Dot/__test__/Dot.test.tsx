import {render} from '@testing-library/react-native';

import {COLORS} from '@/mocks';
import Dot from '..';

describe('Dot Component', () => {
  const onSelect = jest.fn();
  const props = {
    color: COLORS[0],
    hasBorder: false,
    onSelect,
  };
  it('renders correctly with no border', () => {
    const {toJSON} = render(<Dot {...props} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with border', () => {
    const {toJSON} = render(<Dot {...props} hasBorder />);

    expect(toJSON()).toMatchSnapshot();
  });
});
