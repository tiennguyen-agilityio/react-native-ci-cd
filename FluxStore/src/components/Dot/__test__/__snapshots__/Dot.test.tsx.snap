// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dot Component renders correctly with border 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      {
        "justifyContent": "center",
      },
      {
        "alignItems": "center",
      },
      {
        "flexWrap": "nowrap",
      },
      undefined,
      {
        "height": 34,
        "width": 34,
      },
    ]
  }
  testID="flex"
>
  <View
    style={
      [
        {
          "backgroundColor": "black",
          "borderColor": "#fff",
          "borderRadius": "50%",
          "borderWidth": 5,
          "elevation": 4,
          "height": 34,
          "shadowOffset": {
            "height": 3,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 6,
          "width": 34,
        },
        {
          "transform": [
            {
              "scale": 1,
            },
          ],
        },
      ]
    }
    testID="dot-black"
  />
</View>
`;

exports[`Dot Component renders correctly with no border 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      {
        "justifyContent": "center",
      },
      {
        "alignItems": "center",
      },
      {
        "flexWrap": "nowrap",
      },
      undefined,
      {
        "height": 34,
        "width": 34,
      },
    ]
  }
  testID="flex"
>
  <View
    style={
      [
        {
          "backgroundColor": "black",
          "borderRadius": "50%",
          "height": 24,
          "shadowColor": "#000",
          "width": 24,
        },
        {
          "transform": [
            {
              "scale": 1,
            },
          ],
        },
      ]
    }
    testID="dot-black"
  />
</View>
`;
